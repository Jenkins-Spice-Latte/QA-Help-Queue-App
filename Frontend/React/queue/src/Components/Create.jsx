import React, { useState } from "react";
import axios from 'axios';
import { CustomInput, Button, Modal, ModalHeader, ModalBody, ModalFooter, Form, FormGroup, Label, InputGroup, Input } from 'reactstrap';

const Create = (props) => {
  const [modal, setModal] = useState(false);

  const toggle = () => setModal(!modal);
  const [authorSt, setAuthor] = useState('');
  const [completeSt, setComplete] = useState("false");
  const [descriptionSt, setDescription] = useState('');
  const [timeSt, setTime] = useState('');
  const [titleSt, setTitle] = useState('');
  const [topicSt, setTopic] = useState('');
  const [urgencySt, setUrgency] = useState('');
  const date = +new Date;
  let disabled;
  let checkAuth;
  let checkTitle;
  let checkDesc;
  let urgencyCheck;
  let topicCheck;

  if(props.mode === "Trainer mode"){
    disabled = false;
  } else{
    disabled = true;
  }

  if(urgencySt !== ""){
    urgencyCheck = <p id="createUrgencyCheck">Urgency selected</p> 
  }

  if(topicSt !== ""){
    topicCheck = <p id="createTopicCheck">Topic selected</p> 
  }
  
  if(authorSt === ""){
    checkAuth = <Input type="text" name="author" id="author" value={authorSt} onChange={(e) => setAuthor(e.target.value)} placeholder="Author name"/>
    
  } else{
    checkAuth = <Input valid type="text" name="author" id="author" value={authorSt} onChange={(e) => setAuthor(e.target.value)} placeholder="Author name"/>
  }

  if(titleSt === ""){
    checkTitle = <Input type="text" name="title" id="title" value={titleSt} onChange={(e) => setTitle(e.target.value)} placeholder="Ticket title"/>
    
  } else{
    checkTitle = <Input valid type="text" name="title" id="title" value={titleSt} onChange={(e) => setTitle(e.target.value)} placeholder="Ticket title"/>
  }

  if(descriptionSt === ""){
    checkDesc = <Input type="textarea" name="description" value={descriptionSt} onChange={(e) => setDescription(e.target.value)} id="description" placeholder="Description" />
    
  } else{
    checkDesc = <Input valid type="textarea" name="description" value={descriptionSt} onChange={(e) => setDescription(e.target.value)} id="description" placeholder="Description" />
  }

  const isEnabled = authorSt.length > 0 && titleSt.length > 0 && descriptionSt.length > 0;
      


  const handleSubmit = event => {
    event.preventDefault();

    let ticket = {
      author: authorSt,
      complete: completeSt,
      description: descriptionSt,
      time_created: Date.now(),
      title: titleSt,
      topic: topicSt,
      urgency: urgencySt
    };

    axios.post(`api/createTicket/create`,  ticket)
      .then(res => {
        console.log(res);
        props.switchLoaded();
      })
  }

    return (
      <>
        <Button color="success" size="lg" disabled={disabled} onClick={toggle}>Create a ticket</Button>
      <div>
        <Modal isOpen={modal} toggle={toggle}>
        <Form onSubmit={handleSubmit}>
          <ModalHeader toggle={toggle}>Create a ticket</ModalHeader>
          <ModalBody>
          
            <InputGroup>
              {checkAuth}                
            </InputGroup>
            <br />
            <InputGroup>
              {checkTitle}
            </InputGroup>
            <br />
            <InputGroup>
              {checkDesc}
            </InputGroup>
            <br />
          <FormGroup>
            <Label for="radioLabel">Topic</Label>
            <div>
              <CustomInput type="radio" id="topic1" name="topic" onChange={(e) => setTopic(e.target.value)} value="Topic1" label="Topic 1" />
              <CustomInput type="radio" id="topic2" name="topic" onChange={(e) => setTopic(e.target.value)} value="Topic2" label="Topic 2" />
              <CustomInput type="radio" id="topic3" name="topic" onChange={(e) => setTopic(e.target.value)} value="Topic3" label="Topic 3" />
              <CustomInput type="radio" id="topic4" name="topic" onChange={(e) => setTopic(e.target.value)} value="Topic4" label="Topic 4" />
              <CustomInput type="radio" id="topic5" name="topic" onChange={(e) => setTopic(e.target.value)} value="Topic5" label="Topic 5" />
            </div>
            {topicCheck}
          </FormGroup>
          <br />
          <FormGroup>
            <Label for="radioLabel">Urgency</Label>
            <div>
              <CustomInput type="radio" id="urgency1" onChange={(e) => setUrgency(e.target.value)} name="urgency" value="1" label="Most urgent" />
              <CustomInput type="radio" id="urgency2" onChange={(e) => setUrgency(e.target.value)} name="urgency" value="2" label="Very urgent" />
              <CustomInput type="radio" id="urgency3" onChange={(e) => setUrgency(e.target.value)} name="urgency" value="3" label="Slightly urgent" />
              <CustomInput type="radio" id="urgency4" onChange={(e) => setUrgency(e.target.value)} name="urgency" value="4" label="Less urgent" />
              <CustomInput type="radio" id="urgency5" onChange={(e) => setUrgency(e.target.value)} name="urgency" value="5" label="Least urgent" />
            </div>
            {urgencyCheck}
          </FormGroup>
            <br />
            <Input type="hidden" name="time" id="time" value={date} onSubmit={(e) => setTime(e.target.value)}/>
          
          </ModalBody>
          <ModalFooter>
            <Button type="submit" disabled={!isEnabled} color="primary" onClick={toggle}>Create</Button>{' '}
            <Button color="secondary" onClick={toggle}>Cancel</Button>
          </ModalFooter>
          </Form>
        </Modal>
      </div>
      </>
    );
  };
  
  export default Create;