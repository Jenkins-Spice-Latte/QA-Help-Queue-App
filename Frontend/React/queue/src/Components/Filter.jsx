import React from 'react';
import { CustomInput, Form, FormGroup, Label, Input } from 'reactstrap';

const Filter = (props) => {
    return (
      <> 
        <p>Filter</p>
        <Form>
          <FormGroup>
            <Label for="exampleCheckbox">Author</Label>
            <div>
              <Input type="text" name="author" id="author" placeholder="Filter by author" onChange={(e) => props.setAuthorFilter(e.target.value)}/>
            </div>
          </FormGroup>
          <FormGroup>
            <Label for="exampleCheckbox">Urgency</Label>
            <div>
              <CustomInput type="checkbox" id="CustomCheckbox8" onClick={() => props.urgencyCheck(1)} checked={props.urgent.includes(1)} value="1" label="Most urgent" />
              <CustomInput type="checkbox" id="CustomCheckbox9" onClick={() => props.urgencyCheck(2)} checked={props.urgent.includes(2)} value="2" label="Very urgent" />
              <CustomInput type="checkbox" id="CustomCheckbox10"onClick={() => props.urgencyCheck(3)} checked={props.urgent.includes(3)} value="3" label="Slightly urgent" />
              <CustomInput type="checkbox" id="CustomCheckbox11"onClick={() => props.urgencyCheck(4)} checked={props.urgent.includes(4)} value="4" label="Less urgent" />
              <CustomInput type="checkbox" id="CustomCheckbox12"onClick={() => props.urgencyCheck(5)} checked={props.urgent.includes(5)} value="5" label="Least urgent" />
            </div>
          </FormGroup>
          <FormGroup>
            <Label for="exampleCheckbox">Topic</Label>
            <div>
              <CustomInput type="checkbox" id="CustomCheckbox13" onClick={() => props.topicCheck("Dev Ops")} checked={props.topic.includes("Dev Ops")} value="Dev Ops" label="Dev Ops" />
              <CustomInput type="checkbox" id="CustomCheckbox14" onClick={() => props.topicCheck("General")} checked={props.topic.includes("General")} value="General" label="General" />
              <CustomInput type="checkbox" id="CustomCheckbox15" onClick={() => props.topicCheck("Back-end")} checked={props.topic.includes("Back-end")} value="Back-end" label="Back-end" />
              <CustomInput type="checkbox" id="CustomCheckbox16" onClick={() => props.topicCheck("Front-end")} checked={props.topic.includes("Front-end")} value="Front-end" label="Front-end" />
              <CustomInput type="checkbox" id="CustomCheckbox17" onClick={() => props.topicCheck("Software")} checked={props.topic.includes("Software")} value="Software" label="Software" />
            </div>
          </FormGroup>
        </Form>
      </>
    );
  };
  
  export default Filter;