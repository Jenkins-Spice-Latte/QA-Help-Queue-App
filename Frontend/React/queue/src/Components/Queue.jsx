import { useState } from 'react';
import axios from 'axios';
import Ticket from './Ticket'

const Queue = (props) => {
  const {className} = props;

  console.log("queue")
  console.log(props.isLoaded)

  const [data, setData] = useState([]);
  
  let topic1data = [];
  let topic2data = [];
  let topic3data = [];
  let topic4data = [];
  let topic5data = [];

  let urgency1data = [];
  let urgency2data = [];
  let urgency3data = [];
  let urgency4data = [];
  let urgency5data = [];

  if(props.topicfilter.includes("Topic1")) {
    topic1data = data.filter(item => item.topic === "Topic1");
  }

  if(props.topicfilter.includes("Topic2")) {
    topic2data = data.filter(item => item.topic === "Topic2");
  }

  if(props.topicfilter.includes("Topic3")) {
    topic3data = data.filter(item => item.topic === "Topic3");
  }

  if(props.topicfilter.includes("Topic4")) {
    topic4data = data.filter(item => item.topic === "Topic4");
  }

  if(props.topicfilter.includes("Topic5")) {
    topic5data = data.filter(item => item.topic === "Topic5");
  }
  
  let topicdata = topic1data.concat(topic2data, topic3data, topic4data, topic5data);

  if(props.urgentfilter.includes(1)) {
    urgency1data = topicdata.filter(item => item.urgency === 1);
  }

  if(props.urgentfilter.includes(2)) {
    urgency2data = topicdata.filter(item => item.urgency === 2);
  }

  if(props.urgentfilter.includes(3)) {
    urgency3data = topicdata.filter(item => item.urgency === 3);
  }

  if(props.urgentfilter.includes(4)) {
    urgency4data = topicdata.filter(item => item.urgency === 4);
  }

  if(props.urgentfilter.includes(5)) {
    urgency5data = topicdata.filter(item => item.urgency === 5);
  }

  let urgencydata = urgency1data.concat(urgency2data, urgency3data, urgency4data, urgency5data);

  const result = urgencydata.filter(item => 
    item.author.toLowerCase().includes(props.authorfilter.toLowerCase())
  );

  if(props.isLoaded === false){
      axios.get("http://localhost:8902/readAll")
      .then(response => {
          console.log(response.data)
          setData(response.data);
          props.switchLoaded(); 
      });
  }

    return (
      <> 
      <div className= "queue_div">
        <p>Pending Tickets</p>
        {props.sort}
        {result.map((item) => {
            if(item.complete === false)
              return <Ticket item={item} className={className} mode={(props.mode)} switchLoaded={props.switchLoaded} isLoaded={props.isLoaded}/>  
          })}
      </div>
        
      <div className= "queue_div">
        <p>Completed Tickets</p>
        {result.map((item) => {
          if(item.complete === true)
            return <Ticket item={item} className={className} switchLoaded={props.switchLoaded} isLoaded={props.isLoaded}/>  
         })}
      </div>
        
      </>
    );
  };
  
  export default Queue;