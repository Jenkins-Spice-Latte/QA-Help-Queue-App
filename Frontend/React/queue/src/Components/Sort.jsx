import '../App.css';
const Sort = () => {
  return (
    <> 
      <p>Sort</p>
      <p>By newest to oldest</p>
      <p>By oldest to newest</p>
      <p>In ascending order (of title)</p>
    </>
  );
};

export default Sort;