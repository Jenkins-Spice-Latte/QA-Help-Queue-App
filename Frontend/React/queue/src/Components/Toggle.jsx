import React, { useState } from 'react';
import { Button } from 'reactstrap';
import { BsToggleOff,  BsToggleOn} from "react-icons/bs";
import Queue from "./Queue";
import Create from "./Create";

const Toggle = (props) => {
  const [modeSelect, setmodeSelect] = useState("Trainee mode");
  var mode;

  const onCheckboxBtnClick = (selected) => {
    setmodeSelect(selected);
  }

  if(modeSelect === "Trainee mode"){
    mode = <Button color="primary" onClick={() => onCheckboxBtnClick("Trainer mode")}><BsToggleOff className="trainerBtn"/> Use trainer mode</Button>
  } else{
    mode = <Button color="primary" onClick={() => onCheckboxBtnClick("Trainee mode")}><BsToggleOn className="traineeBtn"/> Use trainee mode</Button>    
  } 


    return (
      <> 
        {mode}
        <Queue mode={(modeSelect)}/>
        <Create mode={(modeSelect)}/>
      </>
    );
  };
  
  export default Toggle;