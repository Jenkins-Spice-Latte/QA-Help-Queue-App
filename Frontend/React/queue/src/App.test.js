import renderer from 'react-test-renderer';
import Sort from './Components/Sort';
import Filter from './Components/Filter';
import Create from './Components/Create';
import Queue from './Components/Queue';
import Toggle from './Components/Toggle';

test('Check if modeSelect, sortChange and authorFilter are valid', () => {
  expect(modeSelect).toBeDefined();
  expect(modeSelect).toBeNaN();
  expect(modeSelect).toContain("mode");

  expect(sortChange).toBeDefined();
  expect(sortChange).toBeNaN();

  expect(authorFilter).toBeNaN();
});

test('Check if setModeSelect, setsortChange, setAuthorFilter and setIsLoaded are working', () => {
  setmodeSelect("Trainer mode");
  expect(modeSelect).toEqual("Trainer mode");

  setsortChange("Newest");
  expect(sortChange).toEqual("Newest");
  
  setAuthorFilter("Newest");
  expect(authorFilter).toEqual("Newest");
  
  setIsLoaded(true);
  expect(isLoaded).toEqual(true);
});

test('Check if setUrgentFilter and setTopicFilter are working', () => {
  const urgentexpected = [6];
  const topicexpected = ["Topic7"];
  setUrgentFilter([...6]);
  setTopicFilter([..."Topic7"]);
  
  expect(urgentFilterChange).toEqual(
    expect.arrayContaining(urgentexpected),
  );
  expect(topicFilterChange).toEqual(
    expect.arrayContaining(topicexpected),
  );

});

test('Check if onCheckboxBtnClick, switchLoaded and onSortBtnClick are working', () => {
  setmodeSelect("Trainer mode");
  onCheckboxBtnClick("Trainee mode");
  expect(modeSelect).toEqual("Trainee mode");

  setIsLoaded(true);
  switchLoaded();
  expect(setIsLoaded).toEqual(false);

  setsortChange("Newest");
  onSortBtnClick("Oldest");
  expect(sortChange).toEqual("Oldest");

});


test('Check if onFilterUrgentCheckboxClick and onFilterTopicCheckboxClick are working', () => {
  setUrgentFilter([]);
  onFilterUrgentCheckboxClick([1]);
  expect(urgentFilterChange).toEqual([1]);

  setTopicFilter([]);
  onFilterTopicCheckboxClick(["Topic1"]);
  expect(topicFilterChange).toEqual(["Topic1"]);
});

test('tests to see if Create renders properly', () => {
  const createComp = renderer.create(<Create mode={(modeSelect)} switchLoaded={switchLoaded}/>).toJSON();
  expect(createComp).toMatchSnapshot();
});

test('tests to see if Sort renders properly', () => {
  const SortComp = renderer.create(<Sort onSortBtnClick={onSortBtnClick}/>).toJSON();
  expect(SortComp).toMatchSnapshot();
});

test('tests to see if Filter renders properly', () => {
  const filterComp = renderer.create(<Filter urgencyCheck={onFilterUrgentCheckboxClick} setAuthorFilter={setAuthorFilter} urgent={(urgentFilterChange)} 
  topicCheck={onFilterTopicCheckboxClick} topic={(topicFilterChange)}/>).toJSON();
  expect(filterComp).toMatchSnapshot();
});

test('tests to see if Toggle renders properly', () => {
  const toggleComp = renderer.create(<Toggle onCheckboxBtnClick={onCheckboxBtnClick} mode={modeSelect}/>).toJSON();
  expect(toggleComp).toMatchSnapshot();
});

test('tests to see if Queue renders properly', () => {
  const QueueComp = renderer.create(<Queue mode={(modeSelect)} sort={(sortChange)} urgentfilter={(urgentFilterChange)} 
  topicfilter={(topicFilterChange)} authorfilter={(authorFilter)} switchLoaded={switchLoaded} isLoaded={(isLoaded)}/>).toJSON();
  expect(QueueComp).toMatchSnapshot();
});
