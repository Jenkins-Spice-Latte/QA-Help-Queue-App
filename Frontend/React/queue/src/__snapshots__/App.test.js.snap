// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests to see if Create renders properly 1`] = `
Array [
  <button
    aria-label={null}
    className="btn btn-success btn-lg disabled"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Create a ticket
  </button>,
  <div />,
]
`;

exports[`tests to see if Filter renders properly 1`] = `
Array [
  <p>
    Filter
  </p>,
  <form
    className=""
  >
    <div
      className="form-group"
    >
      <label
        className=""
        htmlFor="exampleCheckbox"
      >
        Author
      </label>
      <div>
        <input
          className="form-control"
          id="author"
          name="author"
          onChange={[Function]}
          placeholder="Filter by author"
          type="text"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
        htmlFor="exampleCheckbox"
      >
        Urgency
      </label>
      <div>
        <div
          className="custom-checkbox custom-control"
          hidden={false}
        >
          <input
            checked={true}
            className="custom-control-input"
            id="CustomCheckbox8"
            onClick={[Function]}
            type="checkbox"
            value="1"
          />
          <label
            className="custom-control-label"
            htmlFor="CustomCheckbox8"
          >
            Most urgent
          </label>
        </div>
        <div
          className="custom-checkbox custom-control"
          hidden={false}
        >
          <input
            checked={true}
            className="custom-control-input"
            id="CustomCheckbox9"
            onClick={[Function]}
            type="checkbox"
            value="2"
          />
          <label
            className="custom-control-label"
            htmlFor="CustomCheckbox9"
          >
            Very urgent
          </label>
        </div>
        <div
          className="custom-checkbox custom-control"
          hidden={false}
        >
          <input
            checked={true}
            className="custom-control-input"
            id="CustomCheckbox10"
            onClick={[Function]}
            type="checkbox"
            value="3"
          />
          <label
            className="custom-control-label"
            htmlFor="CustomCheckbox10"
          >
            Slightly urgent
          </label>
        </div>
        <div
          className="custom-checkbox custom-control"
          hidden={false}
        >
          <input
            checked={true}
            className="custom-control-input"
            id="CustomCheckbox11"
            onClick={[Function]}
            type="checkbox"
            value="4"
          />
          <label
            className="custom-control-label"
            htmlFor="CustomCheckbox11"
          >
            Less urgent
          </label>
        </div>
        <div
          className="custom-checkbox custom-control"
          hidden={false}
        >
          <input
            checked={true}
            className="custom-control-input"
            id="CustomCheckbox12"
            onClick={[Function]}
            type="checkbox"
            value="5"
          />
          <label
            className="custom-control-label"
            htmlFor="CustomCheckbox12"
          >
            Least urgent
          </label>
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
        htmlFor="exampleCheckbox"
      >
        Topic
      </label>
      <div>
        <div
          className="custom-checkbox custom-control"
          hidden={false}
        >
          <input
            checked={true}
            className="custom-control-input"
            id="CustomCheckbox13"
            onClick={[Function]}
            type="checkbox"
            value="Dev Ops"
          />
          <label
            className="custom-control-label"
            htmlFor="CustomCheckbox13"
          >
            Dev Ops
          </label>
        </div>
        <div
          className="custom-checkbox custom-control"
          hidden={false}
        >
          <input
            checked={true}
            className="custom-control-input"
            id="CustomCheckbox14"
            onClick={[Function]}
            type="checkbox"
            value="General"
          />
          <label
            className="custom-control-label"
            htmlFor="CustomCheckbox14"
          >
            General
          </label>
        </div>
        <div
          className="custom-checkbox custom-control"
          hidden={false}
        >
          <input
            checked={true}
            className="custom-control-input"
            id="CustomCheckbox15"
            onClick={[Function]}
            type="checkbox"
            value="Back-end"
          />
          <label
            className="custom-control-label"
            htmlFor="CustomCheckbox15"
          >
            Back-end
          </label>
        </div>
        <div
          className="custom-checkbox custom-control"
          hidden={false}
        >
          <input
            checked={true}
            className="custom-control-input"
            id="CustomCheckbox16"
            onClick={[Function]}
            type="checkbox"
            value="Front-end"
          />
          <label
            className="custom-control-label"
            htmlFor="CustomCheckbox16"
          >
            Front-end
          </label>
        </div>
        <div
          className="custom-checkbox custom-control"
          hidden={false}
        >
          <input
            checked={true}
            className="custom-control-input"
            id="CustomCheckbox17"
            onClick={[Function]}
            type="checkbox"
            value="Software"
          />
          <label
            className="custom-control-label"
            htmlFor="CustomCheckbox17"
          >
            Software
          </label>
        </div>
      </div>
    </div>
  </form>,
]
`;

exports[`tests to see if Queue renders properly 1`] = `
Array [
  <div
    className="queue_div"
  >
    <p
      class="ticket_title"
      id="pending_title"
    >
      Pending Tickets
    </p>
  </div>,
  <div
    className="queue_div"
  >
    <p
      class="ticket_title"
      id="completed_title"
    >
      Completed Tickets
    </p>
  </div>,
]
`;

exports[`tests to see if Sort renders properly 1`] = `
<form
  className=""
>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="exampleCheckbox"
    >
      Search here
    </label>
    <div>
      <input
        className="form-control"
        id="keyword"
        name="keyword"
        onChange={[Function]}
        placeholder="Keyword search"
        type="text"
      />
    </div>
  </div>
</form>
`;

exports[`tests to see if Toggle renders properly 1`] = `
<button
  aria-label={null}
  className="btn btn-primary"
  onClick={[Function]}
  type="button"
>
  <svg
    className="traineeBtn"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 16 16"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M5 3a5 5 0 000 10h6a5 5 0 000-10H5zm6 9a4 4 0 100-8 4 4 0 000 8z"
      fillRule="evenodd"
    />
  </svg>
   Use trainee mode
</button>
`;
